server/src/main/java/org/elasticsearch/index/translog/TranslogHeader.java

```diff
@@ -0,0 +1,164 @@
+            new BufferedChecksumStreamInput(new InputStreamStreamInput(Channels.newInputStream(channel), channel.size()));
+        final int version;
+        try {
+            version = CodecUtil.checkHeader(new InputStreamDataInput(in), TRANSLOG_CODEC, VERSION_CHECKPOINTS, VERSION_PRIMARY_TERM);
+        } catch (CorruptIndexException | IndexFormatTooOldException | IndexFormatTooNewException e) {
+            throw new TranslogCorruptedException("Translog header corrupted. path:" + path, e);
+        }
+        // Read the translogUUID
+        final int uuidLen = in.readInt();
+        if (uuidLen > channel.size()) {
+            throw new TranslogCorruptedException("uuid length can't be larger than the translog");
+        }
+        final BytesRef uuid = new BytesRef(uuidLen);
+        uuid.length = uuidLen;
+        in.read(uuid.bytes, uuid.offset, uuid.length);
+        final BytesRef expectedUUID = new BytesRef(translogUUID);
+        if (uuid.bytesEquals(expectedUUID) == false) {
+            throw new TranslogCorruptedException("expected shard UUID " + expectedUUID + " but got: " + uuid +
+                " this translog file belongs to a different translog. path:" + path);
+        }
+        // Read the primary term
+        final long primaryTerm;
+        if (version == VERSION_PRIMARY_TERM) {
+            primaryTerm = in.readLong();
+            assert primaryTerm >= 0 : "Primary term must be non-negative [" + primaryTerm + "]; translog path [" + path + "]";
+        } else {
+            assert version == VERSION_CHECKPOINTS : "Unknown header version [" + version + "]";
+            primaryTerm = UNKNOWN_PRIMARY_TERM;
+        }
+        // Verify the checksum
+        if (version >= VERSION_PRIMARY_TERM) {
+            Translog.verifyChecksum(in);
+        }
+        final int headerSizeInBytes = headerSize(version, uuid.length);
+        assert channel.position() == headerSizeInBytes :
+            "Header is not fully read; header size [" + headerSizeInBytes + "], position [" + channel.position() + "]";
+        return new TranslogHeader(translogUUID, primaryTerm, headerSizeInBytes);
+    }
+
+    /**
+     * Writes this header with the latest format into the file channel
+     */
+    void write(FileChannel channel) throws IOException {
+        // This output is intentionally not closed because closing it will close the FileChannel.
+        final BufferedChecksumStreamOutput out = new BufferedChecksumStreamOutput(
+            new OutputStreamStreamOutput(Channels.newOutputStream(channel)));
+        CodecUtil.writeHeader(new OutputStreamDataOutput(out), TRANSLOG_CODEC, CURRENT_VERSION);
+        // Write uuid
+        final BytesRef uuid = new BytesRef(translogUUID);
+        out.writeInt(uuid.length);
+        out.writeBytes(uuid.bytes, uuid.offset, uuid.length);
+        // Write primary term
+        out.writeLong(primaryTerm);
+        // Checksum header
+        out.writeInt((int) out.getChecksum());
```

Comment:
I think you should call `out.flush()` here to ensure everything is written to the channel.

Reply:
Done